File: Core/harness/memory/workspaces/_mru.py
"""MRU cap settings — single home for ``MRU_LIMIT`` and related constants.

The MRU cap is persisted to ``workspace_settings.json`` so the user
can tune it from the Settings UI without restarting the harness.
Activation logic in ``_store.py`` calls :func:`get_mru_limit` to read
the effective value; clamping ensures garbage on disk falls back to
the default rather than crashing the registry.
"""

from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any, Dict

from .._common import DATA_DIR, ensure_dir

logger = logging.getLogger("wylde.harness.memory.workspaces")

# MRU cap. Activating one past the cap evicts the oldest. The user can
# tune this from the Settings UI; ``MRU_LIMIT`` is the *default* the
# constant kept for back-compat with callers that didn't get the memo,
# but eviction logic and ``get_mru_limit()`` always read the persisted
# value below.
MRU_LIMIT_DEFAULT: int = 5
MRU_LIMIT_MIN: int = 1
MRU_LIMIT_MAX: int = 20
MRU_LIMIT: int = MRU_LIMIT_DEFAULT  # back-compat shim

SETTINGS_PATH: Path = DATA_DIR / "workspace_settings.json"


def _clamp_mru(value: Any) -> int:
    """Validate the MRU cap. Raises ``ValueError`` on garbage so the
    pipe surface can return a structured ``bad_request``."""
    if isinstance(value, bool):
        # bool is a subclass of int; reject explicitly so True/False
        # don't sneak through as 1/0.
        raise ValueError("mru limit must be an integer, not bool")
    try:
        n = int(value)
    except (TypeError, ValueError):
        raise ValueError(f"mru limit must be an integer, got {value!r}")
    if n < MRU_LIMIT_MIN or n > MRU_LIMIT_MAX:
        raise ValueError(
            f"mru limit must be in [{MRU_LIMIT_MIN}, {MRU_LIMIT_MAX}], got {n}"
        )
    return n


def _read_settings() -> Dict[str, Any]:
    if not SETTINGS_PATH.exists():
        return {}
    try:
        raw = json.loads(SETTINGS_PATH.read_text(encoding="utf-8"))
    except Exception as exc:  # noqa: BLE001
        logger.warning("workspaces: settings unreadable, using defaults: %s", exc)
        return {}
    return raw if isinstance(raw, dict) else {}


def _write_settings(settings: Dict[str, Any]) -> None:
    ensure_dir(SETTINGS_PATH.parent)
    tmp = SETTINGS_PATH.with_suffix(".json.tmp")
    tmp.write_text(json.dumps(settings, indent=2), encoding="utf-8")
    tmp.replace(SETTINGS_PATH)


def get_mru_limit() -> int:
    """Effective MRU cap — reads the persisted value, falls back to
    :data:`MRU_LIMIT_DEFAULT` if no setting is on disk yet."""
    raw = _read_settings().get("mru_limit")
    if raw is None:
        return MRU_LIMIT_DEFAULT
    try:
        return _clamp_mru(raw)
    except ValueError:
        return MRU_LIMIT_DEFAULT


def set_mru_limit(value: Any) -> int:
    """Persist a new MRU cap. If the new cap is smaller than the
    current MRU count, evict the oldest workspaces immediately —
    deleting their index folders but preserving the durable workspace
    memory (same semantics as ``_evict_past_mru``).
    """
    # Lazy import to avoid the _store → _mru → _store cycle at module load.
    from ._store import _evict_past_mru, _load_registry, _registry_lock, _save_registry

    n = _clamp_mru(value)
    with _registry_lock:
        settings = _read_settings()
        settings["mru_limit"] = n
        _write_settings(settings)
        # Apply immediately if the cap shrank below the current count.
        workspaces = _load_registry()
        evicted = _evict_past_mru(workspaces, limit=n)
        if evicted:
            _save_registry(workspaces)
    if evicted:
        logger.info(
            "workspaces: mru_limit set to %d — evicted %d older workspace(s): %s",
            n,
            len(evicted),
            ", ".join(evicted),
        )
    else:
        logger.info("workspaces: mru_limit set to %d", n)
    return n

File: rust/crates/wylde-harness/src/memory/workspaces/mru.rs
//! MRU cap settings. Rust port of
//! `Core/harness/memory/workspaces/_mru.py`.
//!
//! The cap is persisted to `<data_dir>/workspace_settings.json` so the
//! user can tune it from the Settings UI without restarting the
//! harness. Activation reads via [`get_mru_limit`] and evicts based on
//! the effective value.

use std::path::PathBuf;
use std::sync::Mutex;

use serde_json::Value;

use crate::memory::common::{data_dir, ensure_dir};

pub const MRU_LIMIT_DEFAULT: u64 = 5;
pub const MRU_LIMIT_MIN: u64 = 1;
pub const MRU_LIMIT_MAX: u64 = 20;

static SETTINGS_LOCK: Mutex<()> = Mutex::new(());

fn settings_path() -> PathBuf {
    data_dir().join("workspace_settings.json")
}

/// Validate an externally-supplied MRU cap. Returns a structured
/// `Err` so the IPC layer can surface `bad_request` cleanly.
pub fn clamp_mru(value: &Value) -> Result<u64, MruError> {
    // Reject `true`/`false` even though serde_json treats them as ints
    // in some surfaces; matches Python's `isinstance(bool)` reject.
    if value.is_boolean() {
        return Err(MruError::NotInteger {
            display: value.to_string(),
        });
    }
    let n: i64 = match value {
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                i
            } else if let Some(f) = n.as_f64() {
                if f.fract() != 0.0 {
                    return Err(MruError::NotInteger {
                        display: f.to_string(),
                    });
                }
                f as i64
            } else {
                return Err(MruError::NotInteger {
                    display: n.to_string(),
                });
            }
        }
        Value::String(s) => s.trim().parse::<i64>().map_err(|_| MruError::NotInteger {
            display: s.clone(),
        })?,
        other => {
            return Err(MruError::NotInteger {
                display: other.to_string(),
            })
        }
    };
    if n < MRU_LIMIT_MIN as i64 || n > MRU_LIMIT_MAX as i64 {
        return Err(MruError::OutOfRange { n });
    }
    Ok(n as u64)
}

/// Errors from [`clamp_mru`].
#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum MruError {
    #[error("mru limit must be an integer, got {display}")]
    NotInteger { display: String },
    #[error("mru limit must be in [{min}, {max}], got {n}", min = MRU_LIMIT_MIN, max = MRU_LIMIT_MAX)]
    OutOfRange { n: i64 },
}

fn read_settings() -> Value {
    let path = settings_path();
    if !path.exists() {
        return Value::Null;
    }
    let raw = match std::fs::read_to_string(&path) {
        Ok(s) => s,
        Err(_) => return Value::Null,
    };
    serde_json::from_str::<Value>(&raw).unwrap_or(Value::Null)
}

fn write_settings(settings: &Value) -> std::io::Result<()> {
    let path = settings_path();
    if let Some(parent) = path.parent() {
        ensure_dir(parent)?;
    }
    let tmp = path.with_extension("json.tmp");
    std::fs::write(&tmp, serde_json::to_string_pretty(settings).unwrap())?;
    std::fs::rename(&tmp, &path)?;
    Ok(())
}

/// Effective MRU cap. Returns the persisted value if valid, else the
/// default. Garbage on disk silently falls back to default — matches
/// Python.
pub fn get_mru_limit() -> u64 {
    let settings = read_settings();
    let raw = settings.get("mru_limit").cloned().unwrap_or(Value::Null);
    match clamp_mru(&raw) {
        Ok(n) => n,
        Err(_) => MRU_LIMIT_DEFAULT,
    }
}

/// Persist a new MRU cap. If the new cap is smaller than the current
/// MRU count, immediately evicts the tail past `n` (same semantics as
/// `evict_past_mru` — index dirs removed, workspace memory preserved).
///
/// Returns the clamped value or a structured error.
pub fn set_mru_limit(value: &Value) -> Result<u64, MruError> {
    let n = clamp_mru(value)?;
    let _g = SETTINGS_LOCK.lock().unwrap();

    let mut settings = read_settings();
    if !settings.is_object() {
        settings = serde_json::json!({});
    }
    settings["mru_limit"] = Value::from(n);
    let _ = write_settings(&settings);

    // Apply immediately. The store module gates its registry on its
    // own mutex; we don't have to share locks because the only field
    // overlap is the registry file, which we re-read post-write.
    let mut workspaces = super::store::load_registry();
    let evicted = super::store::evict_past_mru(&mut workspaces, n as usize);
    if !evicted.is_empty() {
        let _ = super::store::save_registry(&workspaces);
    }
    Ok(n)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::memory::workspaces::test_support::TestEnv;
    use serde_json::json;

    #[test]
    fn get_mru_limit_falls_back_to_default_when_unset() {
        let _env = TestEnv::new();
        assert_eq!(get_mru_limit(), MRU_LIMIT_DEFAULT);
    }

    #[test]
    fn set_then_get_round_trips_within_range() {
        let _env = TestEnv::new();
        let v = set_mru_limit(&json!(7)).unwrap();
        assert_eq!(v, 7);
        assert_eq!(get_mru_limit(), 7);
    }

    #[test]
    fn clamp_mru_rejects_bool() {
        assert!(matches!(
            clamp_mru(&json!(true)).unwrap_err(),
            MruError::NotInteger { .. }
        ));
        assert!(matches!(
            clamp_mru(&json!(false)).unwrap_err(),
            MruError::NotInteger { .. }
        ));
    }

    #[test]
    fn clamp_mru_rejects_non_integer() {
        assert!(matches!(
            clamp_mru(&json!("abc")).unwrap_err(),
            MruError::NotInteger { .. }
        ));
        assert!(matches!(
            clamp_mru(&json!(1.5)).unwrap_err(),
            MruError::NotInteger { .. }
        ));
    }

    #[test]
    fn clamp_mru_rejects_out_of_range() {
        assert!(matches!(
            clamp_mru(&json!(0)).unwrap_err(),
            MruError::OutOfRange { .. }
        ));
        assert!(matches!(
            clamp_mru(&json!(21)).unwrap_err(),
            MruError::OutOfRange { .. }
        ));
    }

    #[test]
    fn clamp_mru_accepts_string_integer() {
        assert_eq!(clamp_mru(&json!("3")).unwrap(), 3);
    }

    #[test]
    fn set_mru_limit_evicts_workspaces_past_new_cap() {
        let _env = TestEnv::new();
        // Seed registry with 4 workspaces, cap to 2.
        use crate::memory::workspaces::store::{save_registry, Workspace};
        let entries = (0..4)
            .map(|i| Workspace::new(&format!("/tmp/setcap_{i}")))
            .collect::<Vec<_>>();
        save_registry(&entries).unwrap();
        set_mru_limit(&json!(2)).unwrap();
        let ws = crate::memory::workspaces::store::list_workspaces();
        assert_eq!(ws.len(), 2);
    }

    #[test]
    fn settings_garbage_falls_back_to_default() {
        let _env = TestEnv::new();
        if let Some(parent) = settings_path().parent() {
            let _ = std::fs::create_dir_all(parent);
        }
        std::fs::write(settings_path(), "{not json").unwrap();
        assert_eq!(get_mru_limit(), MRU_LIMIT_DEFAULT);
    }
}
